File: leetcode-go/239.go
package main

type Node struct {
	value int
	index int
	next  *Node
	prev  *Node
}

type Deque struct {
	start  *Node
	end    *Node
	length int
}

func (d *Deque) PushFront(value, index int) {
	node := &Node{
		value: value,
		index: index,
	}
	if d.start == nil {
		d.start = node
		d.end = node
		d.length++
		return
	}
	temp := d.start
	temp.prev = node
	node.next = temp
	d.start = node
	d.length++
}

func (d *Deque) PushBack(value, index int) {
	node := &Node{
		value: value,
		index: index,
	}
	if d.end == nil {
		d.start = node
		d.end = node
		d.length++
		return
	}
	temp := d.end
	temp.next = node
	node.prev = temp
	d.end = node
	d.length++
}

func (d *Deque) PopBack() int {
	value := d.end.value
	if d.start == d.end {
		d.start = nil
		d.end = nil
		d.length = 0
		return value
	}

	d.end = d.end.prev
	d.end.next = nil
	d.length--
	return value
}

func (d *Deque) PopFront() int {
	value := d.start.value
	if d.start == d.end {
		d.start = nil
		d.end = nil
		d.length = 0
		return value
	}

	d.start = d.start.next
	d.start.prev = nil
	d.length--
	return value
}

func (d *Deque) PeekBack() int {
	return d.end.value
}

func (d *Deque) PeekFront() int {
	return d.start.value
}

func maxSlidingWindow(nums []int, k int) []int {
	maximums := []int{}
	deque := Deque{}

	for index, num := range nums {
		for value := deque.PeekBack(); value < num; deque.PopBack() {
		}
		for deque.start.index-index >= k {
			deque.PopFront()
		}

		maximums = append(maximums, deque.PeekFront())
	}

	return maximums
}

// func main() {
// 	nums := []int{1, 3, -1, -3, 5, 3, 6, 7}
// 	k := 3
// 	fmt.Println(maxSlidingWindow(nums, k))
// }

File: leetcode-js/239.js

function Node(value, next, prev) {
    return { value, next, prev };
}

class Deque {
    constructor(size) {
        this.front = null;
        this.end = null;
        this.maxSize = size;
        this.length = 0;
    }

    isEmpty() {
        return !this.front;
    }

    peek() {
        return this.front.value;
    }

    back() {
        return this.end.value;
    }

    popFront() {
        const frontValue = this.front.value;
        if (this.length === 1) {
            this.front = null;
            this.end = null;
        }
        else {
            this.front = this.front.next;
            this.front.prev = null;
        }
        this.length--;
        return frontValue;
    }

    popBack() {
        const backValue = this.end.value;
        if (this.length === 1) {
            this.front = null;
            this.end = null;
        }
        else {
            this.end = this.end.prev;
            this.end.next = null;
        }
        this.length--;
        return backValue;
    }

    pushBack(value) {
        if (this.length === this.maxSize) {
            return;
        }

        let node = Node(value, null, null);
        if (this.length === 0) {
            this.front = node;
            this.end = node;
        }
        else {
            node.prev = this.end;
            this.end.next = node;
            this.end = node;
        }
        this.length++;
    }

    pushFront(value) {
        if (this.length === this.maxSize) {
            return;
        }
        let node = Node(value, null, null);
        if (this.length === 0) {
            this.front = node;
            this.end = node;
        }
        else {
            node.next = this.front;
            this.front.prev = node;
            this.front = node;
        }
        this.length++;
    }
}

var maxSlidingWindow = function (nums, k) {
    if (nums.length < k) return [];
    let deq = new Deque(k); // deque with indexes
    let result = [];

    for (let i = 0; i < nums.length; i++) {
        while (!deq.isEmpty() && deq.peek() <= i - k) deq.popFront();

        while (!deq.isEmpty() && nums[deq.back()] <= nums[i]) deq.popBack();

        deq.pushBack(i);

        if (i - k + 1 >= 0) {
            result.push(nums[deq.peek()]);
        }
    }

    return result;
};

let nums = [1, 3, -1, -3, 5, 3, 6, 7];
let k = 3
console.log(maxSlidingWindow(nums, k));